src/main/main.c

```diff
@@ -67,15 +68,18 @@ enum {
 
 /** File descriptor handler struct */
 struct fhs {
+	struct le le; 	     /**< Hash/Delete entry                 */
 	re_sock_t fd;        /**< File Descriptor                   */
 	int flags;           /**< Polling flags (Read, Write, etc.) */
 	fd_h* fh;            /**< Event handler                     */
 	void* arg;           /**< Handler argument                  */
+	bool active;         /**< In-use                            */
```

Comment:
it might be possible to move storage of this struct to the application.
in this case there is no need for array/hash.

Reply:
Maybe I'm not fully unterstand your suggestion, wouldn't this break all applications badly? Handling all the event details is not easy to implement. So we have to implement it twice (retest and baresip)?

I can think of some memory size optimizations for `fhs`:

- instead of `mem_zalloc` we can use `malloc/memset/free` directly this saves 16 byte per fhs entry.
- `struct le` can be replaced by a `void *next` for a simple linked list and custom hash lookup solution. This saves 24 byte per fhs entry and 8 byte per bucket. the fhs pointer could be re-calculated with `offsetof`.

Reply:
if you look at `struct tmr`, the object is stored by the user

the idea was to do something similar, example:

```c
int   fd_listen(struct fhs *fhs, re_sock_t fd, int flags, fd_h *fh, void *arg);
void  fd_close(struct fhs *fhs);
```

Reply:
Ah I see, but I think this is problematic for multithreading. If `fhs` is destroyed by the application while `fd_poll` is waiting a dangling `fhs` is used. But maybe we can simply mem_ref `fhs` to avoid this. I will think about this approach.

Reply:
I just gave this approach a try: https://github.com/baresip/re/pull/805 (not ready yet)